conda/models/prefix_graph.py

```diff
@@ -0,0 +1,387 @@
+        for node in removed_nodes:
+            self._remove_node(node)
+        self._toposort()
+        return removed_nodes
+
+    @property
+    def records(self):
+        return iter(self.graph)
+
+    def prune(self):
+        """Prune back all packages until all child nodes are anchored by a spec.
+
+        Returns:
+            Tuple[PrefixRecord]: The pruned nodes.
+
+        """
+        graph = self.graph
+        spec_matches = self.spec_matches
+        original_order = tuple(self.graph)
+
+        removed_nodes = set()
+        while True:
+            inverted_graph = {
+                node: set(key for key in graph if node in graph[key])
+                for node in graph
+            }
+            prunable_nodes = tuple(node for node, children in iteritems(inverted_graph)
+                                   if not children and node not in spec_matches)
+            if not prunable_nodes:
+                break
+            for node in prunable_nodes:
+                removed_nodes.add(node)
+                self._remove_node(node)
+
+        removed_nodes = tuple(filter(
+            lambda node: node in removed_nodes,
+            original_order
+        ))
+        self._toposort()
+        return removed_nodes
+
+    def get_node_by_name(self, name):
+        return next(rec for rec in self.graph if rec.name == name)
+
+    def all_descendants(self, node):
+        graph = self.graph
+        inverted_graph = {
+            node: set(key for key in graph if node in graph[key])
+            for node in graph
+        }
+
+        nodes = [node]
+        nodes_seen = set()
+        i = 0
+        while i < len(nodes):
+            for child_node in inverted_graph[nodes[i]]:
+                if child_node not in nodes_seen:
+                    nodes_seen.add(child_node)
+                    nodes.append(child_node)
+            i += 1
+        return tuple(
+            filter(
+                lambda node: node in nodes_seen,
+                self.graph
+            )
+        )
+
+    def all_ancestors(self, node):
+        graph = self.graph
+        nodes = [node]
+        nodes_seen = set()
+        i = 0
+        while i < len(nodes):
+            for parent_node in graph[nodes[i]]:
+                if parent_node not in nodes_seen:
+                    nodes_seen.add(parent_node)
+                    nodes.append(parent_node)
+            i += 1
+        return tuple(
+            filter(
+                lambda node: node in nodes_seen,
+                graph
+            )
+        )
+
+    def _remove_node(self, node):
+        """ Removes this node and all edges referencing it. """
+        graph = self.graph
+        if node not in graph:
+            raise KeyError('node %s does not exist' % node)
+        graph.pop(node)
+        self.spec_matches.pop(node, None)
+
+        for node, edges in iteritems(graph):
+            if node in edges:
+                edges.remove(node)
+
+    def _toposort(self):
```

Comment:
This toposort is a pretty strict re-implementation of the toposort we already use.  It's reimplementation because we're using `PackageRecord`s as nodes, rather than simple strings.  For reference, the original toposort is at `conda/common/toposort.py`.